core/src/commonMain/kotlin/org/aquila3d/core/math/Quaternion.kt

```diff
@@ -0,0 +1,324 @@
+package org.aquila3d.core.math
+
+import kotlin.math.acos
+import kotlin.math.cos
+import kotlin.math.sin
+import kotlin.math.sqrt
+
+/**
+ * A Unit [Quaternion] is used to represent a rotation/orientation in 3D space.
+ *
+ * Assuming the [Quaternion]s in question are normalized (unit length), subsequent rotations can be performed by
+ * multiplying quaternions together. If they are not normalized, then the provided [rotate] functions must be used.
+ *
+ * Note that [Quaternion] math is associative but not commutative - i.e.:
+ *
+ * **q**(**p****r**) == (**q****p**)**r** but **q****p** != **p****q**
+ *
+ * It is advantages to represent orientations with [Quaternion]s as they are not subject to Gimble Lock like Euler
+ * rotations, provide smooth interpolation between orientations and reduce the number of computations needed for any
+ * given mutation.
+ */
+open class Quaternion {
```

Comment:
I think this should be a data class which would bring a lot of useful utility methods with it. I don't think much value exists for making the core math models extendable.

Reply:
Open has been removed. Not using data classes due to chaining and mutability needs.